rasterio/_crs.pyx

```diff
@@ -0,0 +1,283 @@
+"""Coordinate reference systems, class and functions.
+"""
+
+import json
+import logging
+
+from rasterio cimport _gdal, _base
+from rasterio.compat import UserDict
+from rasterio.errors import CRSError
+from rasterio.compat import string_types
+
+log = logging.getLogger(__name__)
+
+class CRS(UserDict):
+    """A container class for coordinate reference system info
+
+    PROJ.4 is the law of this land: http://proj.osgeo.org/. But whereas PROJ.4
+    coordinate reference systems are described by strings of parameters such as
+
+        +proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs
+
+    here we use mappings:
+
+        {'proj': 'longlat', 'ellps': 'WGS84', 'datum': 'WGS84', 'no_defs': True}
+
+    One can set/get any PROJ.4 parameter using a dict-like key/value pair on the
+    object. You can instantiate the object by simply passing a dict to the
+    constructor. E.g.
+
+        crs = CRS({'init': 'epsg:3005'})
+
+    """
+    def is_valid(self):
+        """Check if valid geographic or projected coordinate reference system."""
+        return self.is_geographic() or self.is_projected()
+
+    def is_geographic(self):
+        cdef void *osr_crs = _base._osr_from_crs(self)
```

Comment:
Or this could possibly be wrapped in a context manager to do that for you, for this and the following very similar lines.  I haven't done that in Cython yet, so I don't know if it is easy / possible.  Feel free to save that for a later refactor too.